hari_client/utils/upload.py

```
@@ -0,0 +1,202 @@
+    # create a trace_id to track triggered metadata update jobs
+    metadata_rebuild_trace_id = uuid.uuid4()
+    print(f"metadata_rebuild jobs trace_id: {metadata_rebuild_trace_id}")
+    metadata_rebuild_jobs = hari.trigger_dataset_metadata_rebuild_job(
+        dataset_id=dataset_id, trace_id=metadata_rebuild_trace_id, subset_id=subset_id
+    )
+
+    # track the status of all metadata rebuild jobs and wait for them to finish
+    job_statuses = []
+    jobs_are_still_running = True
+    while jobs_are_still_running:
+        jobs = hari.get_processing_jobs(trace_id=metadata_rebuild_trace_id)
+        job_statuses = [(job.process_name, job.status.value, job.id) for job in jobs]
+
+        jobs_are_still_running = any(
+            [
+                status[1]
+                in [
+                    models.ProcessingJobStatus.CREATED,
+                    models.ProcessingJobStatus.RUNNING,
+                ]
+                for status in job_statuses
+            ]
+        )
+        if jobs_are_still_running:
+            print(f"waiting for metadata_rebuild jobs to finish, {job_statuses=}")
+            time.sleep(10)
+
+    print(f"metadata_rebuild jobs finished with status {job_statuses=}")
+
+
+def get_or_create_dataset(
+    hari: HARIClient, dataset_name: str, user_group: str, is_anonymized: bool
+) -> uuid.UUID:
+    """
+    Check if a dataset with the given name exists. If not, create it.
+
+    Args:
+        hari: An instance of HARIClient to interact with the HARI API.
+        dataset_name: The name of the dataset to check or create.
+        user_group: The user group under which the dataset should be created.
+        is_anonymized: Whether the dataset should be created with anonymized data.
+    Returns:
+         uuid: The UUID of the found or created dataset.
+    """
+    datasets = hari.get_datasets()
+    dataset_names = [dataset.name for dataset in datasets]
+
+    assert user_group is not None, "User group is required."
```

Please don't use assert statements in the package source code. We don't control the runtime environment of whichever scripts that might end up using this code. Assert statements can be disabled in the interpreter settings. Rather throw an exception if you want to explicitly handle problematic input values.